feat: allow to pipe child stdout

zkochan opened the pull request:
In order to use this for pnpm, we need the ability to listen for child's output.

I added a new option - `opts.stdio`, which, when specified, is passed to the child process.

I used byline to log every line of the child's output. This might be a bad solution. I am open for alternatives. Maybe we could just return the stdout stream by `lifecycle()`?

zkochan committed:
feat: allow to pipe child stdout

zkochan commented:
@zkat, @mikesherov pls have a look.

With these changes, I use npm-lifecycle now in `pnpm@1.14.6`

mikesherov reviewed (changes requested):
This looks good, but can you please explain the need for byline and could you also add some unit tests for the changes made? I know the library is undertested at the moment but I'd like to move it in the right direction. Thanks for contributing!

zkochan commented:
ok, I will try to rewrite it w/o using byline. I will collect the chunks in pnpm's reporter

mikesherov commented:
I don't mind byline personally. Was literally asking what it does :-)

zkochan commented:
oh, it glues chunks from the stdout stream and creates a new stream that emits output by line

mikesherov commented on index.js:
```diff
@@ -47,7 +48,7 @@ function lifecycle (pkg, stage, wd, opts) {
       if (er) return reject(er)
 
-      if ((wd.indexOf(opts.dir) !== 0 || _incorrectWorkingDirectory(wd, pkg)) &&
+      if (((wd.indexOf(opts.dir) !== 0 && opts.dir !== path.join(wd, 'node_modules')) || _incorrectWorkingDirectory(wd, pkg)) &&
```
Why is this change necessary? `opts.dir` is always the top level node modules directory of the package you're installing, while `wd` can be any directory under that. Does that not work for `pnpm`?

zkochan replied:
but what about when executing scripts of the project? dir will be `project/node_modules`. wd will be `project`

zkochan replied:
for dependencies it sure works. dir is `project/node_modules`, wd is `project/node_modules/dependency`

zkat replied:
npm calculates the prefix itself. `wd` works out to be the appropriate local prefix.

zkochan replied:
I don't understand, npm doesn't use `@npm/lifecycle` for the project's scripts? I had issues in pnpm when using `@npm/lifecycle` because the value of dir was incorrect when running scripts in `project/package.json`

mikesherov replied:
Can you point me to a WIP branch on pnpm so I can help troubleshoot?

zkochan replied:
There is no WIP branch currently. I did a fork of @npm/lifecycle with these and the [node-gyp](https://github.com/npm/lifecycle/pull/5) change. I revert this part for now and later when all the other things work, I will retry and create a new issue.

(pnpm uses that fork currently and works fine)

zkat replied:
@zkochan `wd` is not the application's working directory: https://github.com/zkat/cipm/blob/latest/index.js#L149

We never give it something inside of the root of the project.

Furthermore, at the *root*, we use basically this code: https://github.com/zkat/cipm/blob/latest/lib/get-prefix.js to calculate what the "package root" is, even if you're, say, inside `lib`. Here's where we pass the prefix down: https://github.com/zkat/cipm/blob/latest/index.js#L36-L41

Using cipm 'cause it's a smaller codebase :)

zkat commented:
So, I'd really like this patch or something similar, too: I'd like to be able to redirect/muffle stdout for cipm.

mikesherov commented:
Happy to land the stdout stuff without the wd check changes if a unit test comes with it. @zkat feel like whipping one up?

zkochan commented:
I added tests but I added them on the pnpm level, I can duplicate them here

mikesherov commented:
Yeah, please duplicate!

zkochan commented:
Thanks, @zkat for the explanation. I will look into that later. For now, I reverted the wd change.

I united the two commits.
Tests added.
Also, converted the data, returned by byline to string, as it returns bytes.

Let me know if further changes are needed.

mikesherov reviewed (approved):
LGTM. Feel free to land @zkat. Thanks again for the contribution!

mikesherov commented:
Thanks again @zkochan

iarna commented on index.js:
```diff
@@ -282,6 +283,12 @@ function runCmd_ (cmd, pkg, env, wd, opts, stage, unsafe, uid, gid, cb_) {
     }
     procError(er)
   })
+  byline(proc.stdout).on('data', function (data) {
```
In the interests of not making a change that we'll need to then break soon after…

What we've been intending to do with npm is to buffer the output from lifecycle scripts and…

1. If there's an error, output at warn level.
2. If there's not an error, output at suppressed by default level (possibly lower than verbose…)

I suppose we _can_ support that later by adding another option that provides a function that gets the proc object and can do with stdout/stderr as it likes (as an alternative to this).

iarna reviewed (approved):
Sorry about taking so long to get to this, it's been a hectic autumn.

As I think we can build on this in a way that won't require breaking changes to `npm-lifecycle` to support npm's design goals, I'm ok with it as it is.
